coordinator/shard_mapper.go

```diff
@@ -0,0 +1,181 @@
+package coordinator
+
+import (
+	"io"
+	"time"
+
+	"github.com/influxdata/influxdb/influxql"
+	"github.com/influxdata/influxdb/services/meta"
+	"github.com/influxdata/influxdb/tsdb"
+)
+
+// IteratorCreator is an interface that combines mapping fields and creating iterators.
+type IteratorCreator interface {
+	influxql.IteratorCreator
+	influxql.FieldMapper
+	io.Closer
+}
+
+// ShardMapper retrieves and maps shards into an IteratorCreator that can later be
+// used for executing queries.
+type ShardMapper interface {
+	MapShards(sources influxql.Sources, opt *influxql.SelectOptions) (IteratorCreator, error)
+}
```

Comment:
Yes, I think. The original intention of this is when I took it from lazy iterator evaluation. It was to allow the ordering of the shards. Here, it's used to organize the shards based on their sources and create a special `IteratorCreator` that can use `FieldDimensions` rather than requiring `FieldDimensions` and the sources to need to be propagated everywhere.

So it maps the shards needed to the sources requested. Eventually, it will probably also implement lazy iterators if we can ever make that performant.